File: Lydian.Unity.Automapper/Mapper.cs
using Lydian.Unity.Automapper.Core;
using Lydian.Unity.Automapper.Core.Handling;
using Microsoft.Practices.Unity;
using System;
using System.Diagnostics.CodeAnalysis;
using System.Diagnostics.Contracts;

namespace Lydian.Unity.Automapper
{
    /// <summary>
    /// Contains extension method entry points to the Unity Auto-mapper.
    /// </summary>
    public static class Mapper
	{
		private readonly static MappingOptions NULL_MAPPING_OPTIONS = new MappingOptions();

		/// <summary>
		/// Automatically maps and registers interfaces and concrete types into the Unity container.
		/// </summary>
		/// <param name="container">The container to be configured.</param>
		/// <param name="types">The array of interfaces and concretes to map up and register.</param>
		public static void AutomapTypes(this IUnityContainer container, params Type[] types)
		{
			Contract.Requires(container != null, "container");
			Contract.Requires(types != null, "types");

			container.AutomapTypes(NULL_MAPPING_OPTIONS, types);
		}

		/// <summary>
		/// Automatically maps and registers interfaces and concrete types into the Unity container.
		/// </summary>
		/// <param name="container">The container to be configured.</param>
		/// <param name="options">Any custom options to use as guidance when mapping.</param>
		/// <param name="types">The array of interfaces and concretes to map up and register.</param>
		[SuppressMessage("Microsoft.Design", "CA1062:Validate arguments of public methods", Justification = "Code Contracts", MessageId = "1")]
		public static void AutomapTypes(this IUnityContainer container, MappingOptions options, params Type[] types)
		{
			Contract.Requires(container != null, "container");
			Contract.Requires(options != null, "options");
			Contract.Requires(types != null, "types");

			var controller = CreateController(container);
			controller.RegisterTypes(options.Behaviors, types);
		}

		/// <summary>
		/// Automatically maps and registers interfaces and concrete types found in the supplied assemblies into the Unity container.
		/// </summary>
		/// <param name="container">The container to be configured.</param>
		/// <param name="assemblyNames">The list of full assembly names to register types from.</param>
		public static void AutomapAssemblies(this IUnityContainer container, params String[] assemblyNames)
		{
			Contract.Requires(container != null, "container");
			Contract.Requires(assemblyNames != null, "assemblyNames");

			container.AutomapAssemblies(NULL_MAPPING_OPTIONS, assemblyNames);
		}

		/// <summary>
		/// Automatically maps and registers interfaces and concrete types found in the supplied assemblies into the Unity container.
		/// </summary>
		/// <param name="container">The container to be configured.</param>
		/// <param name="options">Any custom options to use as guidance when mapping.</param>
		/// <param name="assemblyNames">The list of full assembly names to register types from.</param>
		[SuppressMessage("Microsoft.Design", "CA1062:Validate arguments of public methods", Justification = "Code Contracts", MessageId = "1")]
		public static void AutomapAssemblies(this IUnityContainer container, MappingOptions options, params String[] assemblyNames)
		{
			Contract.Requires(container != null, "container");
			Contract.Requires(options != null, "options");
			Contract.Requires(assemblyNames != null, "assemblyNames");

			var controller = CreateController(container);
			controller.RegisterAssemblies(options.Behaviors, assemblyNames);
		}

		private static MappingController CreateController(IUnityContainer target)
		{
			var internalContainer = new UnityContainer();

			internalContainer.RegisterType<ITypeMappingHandler, TypeMappingHandler>();
			internalContainer.RegisterType<IRegistrationNameFactory, RegistrationNameFactory>();
			internalContainer.RegisterType<ITypeMappingValidator, TypeMappingValidator>();
			internalContainer.RegisterType<IConfigLifetimeManagerFactory, ConfigLifetimeManagerFactory>();
			internalContainer.RegisterType<IInjectionMemberFactory, InjectionMemberFactory>();

			return new MappingController(target, new TypeMappingFactory(), internalContainer);
		}
	}
}

File: Lydian.Unity.Automapper.Fs/Mapper.fs
namespace Lydian.Unity.Automapper

open Lydian.Unity.Automapper.Core.ConfigurationBuilder
open Lydian.Unity.Automapper.Core.ContainerRegistrar
open Lydian.Unity.Automapper.Core.TypeMappingFactory
open Microsoft.Practices.Unity
open System
open System.Reflection
open System.Runtime.CompilerServices

[<Extension>]
/// <summary>
/// Contains extension method entry points to the Unity Automapper.
/// </summary>
type Mapper() = 
    
    static member private registerTypes types (behaviors : MappingBehaviors) container = 
        let configuration = buildConfiguration (types)
        let mappings = createMappings (behaviors, configuration, types)
        registerMappings (container, mappings, configuration, behaviors)
    
    static member private registerAssemblies (assemblyNames : string seq) = 
        let types = 
            assemblyNames
            |> Seq.map Assembly.Load
            |> Seq.collect (fun asm -> asm.GetTypes())
            |> Seq.cache
        Mapper.registerTypes types
    
    static member private NULL_MAPPING_OPTIONS = { Behaviors = MappingBehaviors.None }
    
    /// <summary>
    /// Automatically maps and registers interfaces and concrete types into the Unity container.
    /// </summary>
    /// <param name="container">The container to be configured.</param>
    /// <param name="options">Any custom options to use as guidance when mapping.</param>
    /// <param name="types">The array of interfaces and concretes to map up and register.</param>
    [<Extension>]
    static member AutomapTypes((container : IUnityContainer), options, [<ParamArray>] types : Type []) = 
        container |> Mapper.registerTypes types options.Behaviors
    
    /// <summary>
    /// Automatically maps and registers interfaces and concrete types into the Unity container.
    /// </summary>
    /// <param name="container">The container to be configured.</param>
    /// <param name="types">The array of interfaces and concretes to map up and register.</param>
    [<Extension>]
    static member AutomapTypes((container : IUnityContainer), [<ParamArray>] types : Type []) = 
        Mapper.AutomapTypes(container, Mapper.NULL_MAPPING_OPTIONS, types)
    
    /// <summary>
    /// Automatically maps and registers interfaces and concrete types found in the supplied assemblies into the Unity container.
    /// </summary>
    /// <param name="container">The container to be configured.</param>
    /// <param name="options">Any custom options to use as guidance when mapping.</param>
    /// <param name="assemblyNames">The list of full assembly names to register types from.</param>
    [<Extension>]
    static member AutomapAssemblies((container : IUnityContainer), options, [<ParamArray>] assemblyNames : string []) = 
        container |> Mapper.registerAssemblies assemblyNames options.Behaviors
    
    /// <summary>
    /// Automatically maps and registers interfaces and concrete types found in the supplied assemblies into the Unity container.
    /// </summary>
    /// <param name="container">The container to be configured.</param>
    /// <param name="assemblyNames">The list of full assembly names to register types from.</param>
    [<Extension>]
    static member AutomapAssemblies((container : IUnityContainer), assemblyNames) = 
        Mapper.AutomapAssemblies(container, Mapper.NULL_MAPPING_OPTIONS, assemblyNames)
